src/video/vid_riva128.c

```diff
@@ -0,0 +1,1072 @@
+}
+
+void
+riva128_ptimer_tick(void *p)
+{
+	riva128_t *riva128 = (riva128_t *)p;
+	//nv_riva_log("RIVA 128 PTIMER tick!\n");
+
+	double time = ((double)riva128->ptimer.clock_mul * 500000000.0f) / (double)riva128->ptimer.clock_div;
+	uint32_t tmp;
+	int alarm_check;
+
+	//if(cs == 0x0008 && !riva128->pgraph.beta) nv_riva_log("RIVA 128 PTIMER time elapsed %f alarm %08x, time_low %08x\n", time, riva128->ptimer.alarm, riva128->ptimer.time & 0xffffffff);
+
+	tmp = riva128->ptimer.time;
+	riva128->ptimer.time += (uint64_t)time;
+
+	alarm_check = (riva128->ptimer.alarm - tmp) < (uint32_t)riva128->ptimer.time;
+
+	if(alarm_check && (riva128->ptimer.intr_en & 1))
+	{
+		//nv_riva_log("RIVA 128 PTIMER ALARM interrupt fired!\n");
+		riva128_ptimer_interrupt(0, riva128);
+	}
+}
+
+void
+riva128_mclk_poll(void *p)
+{
+	riva128_t *riva128 = (riva128_t *)p;
+
+	if(riva128->pmc.enable & (1 << 16)) riva128_ptimer_tick(riva128);
+
+	timer_advance_u64(&riva128->mtimer, riva128->mtime);
+}
+
+uint32_t
+riva128_mmio_read_l(uint32_t addr, void *p)
+{
+    riva128_t *riva128 = (riva128_t *)p;
+
+    addr &= 0xffffff;
+
+	pclog("[RIVA 128] MMIO read %08x\n", addr);
+
+	if ((addr >= 0x000000) && (addr <= 0x000fff)) return riva128_pmc_read(addr, riva128);
+	if ((addr >= 0x009000) && (addr <= 0x009fff)) return riva128_ptimer_read(addr, riva128);
+	if ((addr >= 0x100000) && (addr <= 0x100fff)) return riva128_pfb_read(addr, riva128);
+	if ((addr >= 0x680000) && (addr <= 0x680fff)) return riva128_pramdac_read(addr, riva128);
+    if ((addr >= 0x110000) && (addr <= 0x11ffff)) return ((uint32_t *) riva128->bios_rom.rom)[(addr & riva128->bios_rom.mask) >> 2];
+
+    if ((addr >= 0x1800) && (addr <= 0x18ff))
+	return (riva128_pci_read(0,(addr+0) & 0xff,p) << 0) | (riva128_pci_read(0,(addr+1) & 0xff,p) << 8) | (riva128_pci_read(0,(addr+2) & 0xff,p) << 16) | (riva128_pci_read(0,(addr+3) & 0xff,p) << 24);
+
+    switch(addr) {
+	case 0x6013b4: case 0x6013b5:
+	case 0x6013d4: case 0x6013d5:
+	case 0x6013da:
+	case 0x0c03c2: case 0x0c03c3: case 0x0c03c4: case 0x0c03c5:
+	case 0x6813c6: case 0x6813c7: case 0x6813c8: case 0x6813c9: case 0x6813ca: case 0x6813cb: case 0x6813cc:
+		return (riva128_in((addr+0) & 0x3ff,p) << 0) | (riva128_in((addr+1) & 0x3ff,p) << 8) | (riva128_in((addr+2) & 0x3ff,p) << 16) | (riva128_in((addr+3) & 0x3ff,p) << 24);
+		break;
+    }
+
+    return 0;
+}
+
+
+uint8_t
+riva128_mmio_read(uint32_t addr, void *p)
+{
+    riva128_t *riva128 = (riva128_t *)p;
+
+    addr &= 0xffffff;
+
+    if ((addr >= 0x110000) && (addr <= 0x11ffff)) return riva128->bios_rom.rom[addr & riva128->bios_rom.mask];
+
+    if ((addr >= 0x1800) && (addr <= 0x18ff))
+	return riva128_pci_read(0,addr & 0xff,p);
+
+    switch(addr) {
+	case 0x6013b4: case 0x6013b5:
+	case 0x6013d4: case 0x6013d5:
+	case 0x6013da:
+	case 0x0c03c2: case 0x0c03c3: case 0x0c03c4: case 0x0c03c5:
+	case 0x6813c6: case 0x6813c7: case 0x6813c8: case 0x6813c9: case 0x6813ca: case 0x6813cb: case 0x6813cc:

```

Comment:
IMHO multiple lines for the case labels would be clearer but that's just my opinion